_drafts/test-automation-renode.md

```diff
@@ -0,0 +1,886 @@
+your test suite.
+
+```robot
+*** Settings ***
+Library           DateTime
+```
+
+### UART Timeouts
+
+To keep tests failing quickly, make sure not to forget adding
+`timeout=<seconds>` to UART expectations.
+
+```
+    Wait For Line On Uart       help: Lists all commands      timeout=2
+```
+
+If you are building Renode from source or have a release greater than 1.9.0, you
+can also set this globally by overriding the `${DEFAULT_UART_TIMEOUT}` variable
+in your suite.
+
+```robot
+*** Variables ***
+${DEFAULT_UART_TIMEOUT}     2
+```
+
+### Regular Expressions on UART
+
+You can expect and validate strings from the UART by using regular expressions,
+which allows developers to read in values that change between invocations of the
+test. For example, we can test the following shell command:
+
+```c
+int cli_command_greet(int argc, char *argv[]) {
+    char buf[64];
+    snprintf(buf, sizeof(buf), "Hello %s!", argv[1]);
+    shell_put_line(buf);
+    return 0;
+}
+```
+
+By using `treatAsRegex=True` and saving this value to `p`, we can then read it
+out by using `p.groups[0]` and comparing it to the name provided in the test.
+
+```robot
+*** Test Cases ***
+Greet
+    Start Test
+
+    Wait For Prompt On Uart         ${SHELL_PROMPT}
+    Write Line To Uart              greet Tyler
+
+    ${p}=  Wait For Line On Uart    Hello (\\w+)!     treatAsRegex=true     timeout=2
+    Should Be True                  'Tyler' == """${p.groups[0]}"""
+```
+
+### Comparison Operations
+
+At Pebble, we used dynamic memory allocations. To help us sleep at night and
+confirm that we always had enough heap available on every build we shipped, we
+added tests ensuring our high-water-marks were within acceptable limits. A test
+we might have written at the time would measure the available heap space after
+all of the tests had finished.
+
+```robot
+*** Test Cases ***
+High Water Mark
+    Start Test
+
+    Wait For Prompt On Uart         ${SHELL_PROMPT}
+    Write Line To Uart              heap_free
+
+    ${p}=  Wait For Line On Uart    (\\d+)     treatAsRegex=true  timeout=2
+    ${i}=  Convert To Integer       ${p.groups[0]}
+
+    # Ensure we have 5000 bytes free at the end of the tests
+    Should Be True                  5000 < ${i}
+```
+
+In our automated tests, we did similar checks to make sure that our performance,
+power usage, stack, and heap usage were always within acceptable limits.
+
+### Line Endings
+
+If your shell uses carriage returns as line endings, make sure to add the
+following option when creating the tester:
+
+```
+    Create Terminal Tester    ${UART}  endLineOption=TreatCarriageReturnAsEndLine
+```
+
+### Tags & Documentation
+
+You can add a **Tags** and **Documentation** attribute to each test, which will
+help bucket your tests up into logical groups and help developers know what each
+test is doing.
+
+```robot
+*** Test Cases ***
+Command
+    [Documentation]             Some command
```

Comment:
Maybe a screenshot of where the  "Documentation" shows up? Don't believe I see anything below